kotlin/perform-valve-isolation-trace/src/main/java/com/esri/arcgisruntime/sample/performvalveisolationtrace/MainActivity.kt

```diff
@@ -0,0 +1,296 @@
+                }
+              } else {
+                Toast.makeText(
+                  this,
+                  "Error: Starting location geometry must be point.",
+                  Toast.LENGTH_LONG
+                ).show()
+              }
+            } else {
+              Toast.makeText(
+                this,
+                "Error: Starting location features not found.",
+                Toast.LENGTH_LONG
+              ).show()
+            }
+          } catch (e: Exception) {
+            val message = "Error loading starting location feature: ${e.message}"
+            Log.e(TAG, message)
+            Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+          }
+        }
+      } else {
+        Toast.makeText(this, "Error loading utility network.", Toast.LENGTH_LONG).show()
+      }
+    }
+  }
+
+  /**
+   * Performs a valve isolation trace according to the defined trace configuration and starting location.
+   * Selects the resulting features on the map.
+   *
+   * @param utilityNetwork the utility network to perform the trace on
+   * @param traceConfiguration the trace configuration to apply to the trace
+   * @param startingLocation the utility element to begin the trace from
+   */
+  private fun performTrace(
+    utilityNetwork: UtilityNetwork,
+    traceConfiguration: UtilityTraceConfiguration,
+    startingLocation: UtilityElement
+  ) {
+    progressBar.visibility = View.VISIBLE
+    // create a category comparison for the trace
+    // NOTE: UtilityNetworkAttributeComparison or UtilityCategoryComparison with Operator.DoesNotExists
+    // can also be used. These conditions can be joined with either UtilityTraceOrCondition or UtilityTraceAndCondition.
+    val categoryComparison = UtilityCategoryComparison(
+      spinner.selectedItem as UtilityCategory,
+      UtilityCategoryComparisonOperator.EXISTS
+    )
+    // set the category comparison to the barriers of the configuration's trace filter
+    traceConfiguration.filter.barriers = categoryComparison
+
+    // set the configuration to include or leave out isolated features
+    traceConfiguration.isIncludeIsolatedFeatures = include_isolated_switch.isChecked
+
+    // build parameters for the isolation trace
+    val traceParameters =
+      UtilityTraceParameters(UtilityTraceType.ISOLATION, listOf(startingLocation))
+    traceParameters.traceConfiguration = traceConfiguration
+
+    // run the trace and get the result
+    val utilityTraceResultsFuture = utilityNetwork.traceAsync(traceParameters);
+    utilityTraceResultsFuture.addDoneListener {
+      try {
+        (utilityTraceResultsFuture.get()[0] as? UtilityElementTraceResult)?.let { utilityElementTraceResult ->
+          if (utilityElementTraceResult.elements.isNotEmpty()) {
+            Log.i(
+              TAG,
+              "I think the trace was successful ${utilityElementTraceResult.elements.size}"
+            )
```

Comment:
I assume this line isn't meant to be left in :P

Reply:
you would be right hahaha